File: Gravical.Poker.Core/TableFinal.cs
namespace Gravical.Poker.Core
{
    /// <summary>
    /// Holds all community cards from a Texas Hold-em table.
    /// </summary>
    public class TableFinal
    {
        public Card First { get; }
        public Card Second { get; }
        public Card Third { get; }
        public Card Turn { get; }
        public Card River { get; }

        public TableFinal(Card first, Card second, Card third, Card turn, Card river)
        {
            First = first;
            Second = second;
            Third = third;
            Turn = turn;
            River = river;
        }

        public Card[] ToArray()
        {
            return new[] { First, Second, Third, Turn, River };
        }

        public Card[] GetAllCards(Pocket pocket)
        {
            return new[]
            {
                First,
                Second,
                Third,
                Turn,
                River,
                pocket.CardA,
                pocket.CardB,
            };
        }

        public bool Contains(Card card)
        {
            return card.Equals(First)
                   || card.Equals(Second)
                   || card.Equals(Third)
                   || card.Equals(Turn)
                   || card.Equals(River);
        }

        public override string ToString() => ToArray().ToCardString();
    }
}

File: JavaScript/src/TableFinal.mjs
import { Card } from './Card.mjs';
import { Pocket } from './Pocket.mjs';

// Holds all community cards from a Texas Hold-em table.
export class TableFinal {

	constructor() {
		this.first = null;
		this.second = null;
		this.third = null;
		this.turn = null;
		this.river = null;
	}

	static make(first, second, third, turn, river) {
		let table = new TableFinal();
		table.first = first;
		table.second = second;
		table.third = third;
		table.turn = turn;
		table.river = river;
		return table;
	}

	toArray() {
		return [ this.first, this.second, this.third, this.turn, this.river ];
	}

	getAllCards(pocket) {
		if (!pocket || !(pocket instanceof Pocket)) throw "Pocket isn't valid";
		return [
			this.first,
			this.second,
			this.third,
			this.turn,
			this.river,
			pocket.a,
			pocket.b,
		];
	}

	contains(card) {
		if (!card || !(card instanceof Card)) return false;
		return card.equals(this.first)
				|| card.equals(this.second)
				|| card.equals(this.third)
				|| card.equals(this.turn)
				|| card.equals(this.river);
	}

	toString() {
		return `${this.first} ${this.second} ${this.third} ${this.turn} ${this.river}`;
	}
}
